### PR TITLE
Fix the RuntimeJobDag issue for queues

### LINKED ISSUE #1005: RunTimeJobDag might have multiple instances of a job
It has been observed that if ParallelJobs has a value more than one, If controller switch happens or several running jobs finish in the same pipeline, the _readyJobList in the RunTimeJobDag will have multiple instance of the same job. This will cause problems for the queue that has ParallelJobs set to a value grater than one. It means that after controller switch, the jobs will run one by one. We need to add a check to make sure _readyJobList contains unique jobs.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #1005 

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
In this PR, a fix has been implemented to avoid _readyJobList in RuntimeJobDag to contain multiple entries of a job. In this RuntimeJobDah.finishJob we add the next job without considering of it has already been added or not. In this PR, added a test which proves this can happen and I proposed a fix for this issue.

### Tests
- [x] The following tests are written for this issue:
TestEnqueueJobs.testQueueParallelJobs

- [x] The following is the result of the "mvn test" command on the appropriate module:

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestRecurringJobQueue.deleteRecreateRecurrentQueue » ThreadTimeout Method org....
[ERROR]   TestWorkflowTermination.testWorkflowJobFail:223 » Helix Workflow "testWorkflow...
[INFO] 
[ERROR] Tests run: 1145, Failures: 2, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:24 h
[INFO] Finished at: 2020-05-13T13:55:21-07:00
[INFO] ------------------------------------------------------------------------
```

The failed tests have passed when I ran them individually.
```
mvn test -Dtest="TestRecurringJobQueue,TestWorkflowTermination"
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 137.835 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 11, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  02:23 min
[INFO] Finished at: 2020-05-13T14:21:58-07:00
[INFO] ------------------------------------------------------------------------
```

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java [hunk 1]
```diff
@@ -146,8 +146,13 @@ public boolean finishJob(String job) {
     }
     // Add finished job's successors to ready-list
     if (_isJobQueue) {
-      if (_lastJob != null && _parentsToChildren.containsKey(_lastJob)) {
-        _readyJobList.offer(_parentsToChildren.get(_lastJob).iterator().next());
+      while (_lastJob != null && _parentsToChildren.containsKey(_lastJob)) {
```
Comment (dasahcc) [inline helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:149]: Better to have some comments here. I think this is an optimization instead of fix. Even if you add the scheduled one as next job. It will be skipped in schedule job part in WorkflowDispatcher. 

Would that cause problem for get another next job?

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:149]: I will add the comments. Actually this is a fix. If you look at the test I added, it is failing without this fix. If you run the newly added test, you will notice that the jobs are not running in parallel because of this issue.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java [hunk 2]
```diff
@@ -146,8 +146,13 @@ public boolean finishJob(String job) {
...
+        String nextJob = _parentsToChildren.get(_lastJob).iterator().next();
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:150]: Could nextJob be null?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java [hunk 3]
```diff
@@ -96,4 +102,60 @@ public void testJobSubmitGenericWorkflows() throws InterruptedException {
...
+    WorkflowConfig.Builder workflowCfgBuilder = new WorkflowConfig.Builder()
+        .setWorkflowId(queueName).setParallelJobs(3).setAllowOverlapJobAssignment(true);
+    _driver.start(builder.setWorkflowConfig(workflowCfgBuilder.build()).build());
+    JobConfig.Builder jobBuilder =
+        new JobConfig.Builder().setTargetResource(WorkflowGenerator.DEFAULT_TGT_DB)
+            .setCommand(MockTask.TASK_COMMAND).setMaxAttemptsPerTask(2)
+            .setJobCommandConfigMap(Collections.singletonMap(MockTask.JOB_DELAY, "10000"));
+
+    // Add 4 jobs to the queue
+    for (int i = 0; i <= 3; i++) {
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java:119]: Let's make these magic numbers descriptive variables?

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 1 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:149]: +1. I believe we're making this improvement because we've seen an issue in production. Could you add a block comment here what the issue was and why this would fix the issue to give more context?

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java [hunk 4]
```diff
@@ -96,4 +102,60 @@ public void testJobSubmitGenericWorkflows() throws InterruptedException {
...
+    Set<Long> startTimes = new HashSet<>();
+    Set<Long> endTime = new HashSet<>();
+
+    for (int i = 4; i <= 6; i++) {
+      JobContext jobContext =
+          _driver.getJobContext(TaskUtil.getNamespacedJobName(queueName, "JOB" + i));
+      startTimes.add(jobContext.getStartTime());
+      endTime.add(jobContext.getFinishTime());
+    }
+    Assert.assertTrue(Collections.min(endTime) > Collections.max(startTimes));
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java:159]: Could you help understand: If you are only checking one: maxEndtime > maxStartime, why are HashSets necessary?

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java [hunk 5]
```diff
@@ -146,8 +146,13 @@ public boolean finishJob(String job) {
...
+        if (!_readyJobList.contains(nextJob)) {
```
Comment (pkuwm) [inline helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:151]: What's the time complexity of this operation?

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 1 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:149]: Sure. Added comments about it. I will change the PR for master branch.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 2 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:150]: Since _parentsToChildren.containsKey(_lastJob) is only added if there is actually a children for that job, as long as we are in the while loop, it means that nextJob is not null.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java (hunk 3 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java:119]: Done.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 5 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:151]: I understand you concern about time complexity. However, if you look at the whole structure of the code, finish job is only called once and only if the job is finished. Also parallel jobs is not usually set to a large value, I think we are safe here. Also note that we need that for correctness of TF.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java (hunk 4 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java:159]: Changed it to one long and get minimum/maximum in every loop.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 2 above)
Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:157]: could you add that explanation in the code?

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java [hunk 6]
```diff
@@ -145,9 +145,21 @@ public boolean finishJob(String job) {
           String.format("Job: %s has either finished already, never been scheduled, or been removed from DAG", job));
     }
     // Add finished job's successors to ready-list
+
+    // If it is a jobQueue, there should be a check to make sure that the a job has not been added
+    // to the _readyJobList multiple times. This check is necessary because once the controller
+    // switch happens, the _readyJobList and _inflightJobList will be created from scratch. In this
+    // case, since there might be many jobs that have been finished before, we do not want to have a
+    // job several times to the _readyJobList. If _readyJobList has multiple instances of the same
+    // job, it can compromise the functionality of the parallel jobs.
```
Comment (narendly) [inline helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:154]: Have a follow up question. What do you mean the job gets added several times to the readyJobList?

If there's a controller switch, we'll be creating readyJobList/inflightJobList from scratch, and add everything from the beginning (or the first available job from the step-by-step topological sort). Then as long as the controller doesn't switch, it only gets added once, right?

In other words, for each controller should add a job exactly once. Is that correct?

Also even though a job might get added to readyJobList multiple times (exactly once in different controllers), our scheduling logic just skips it I believe (if jobState is complete/progress we just skip?). Could you help me understand how this could help? It seems that although this wouldn't hurt to do...

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java [hunk 7]
```diff
@@ -1,12 +1,18 @@
 package org.apache.helix.integration.task;
 
+import java.util.Collections;
+import java.util.HashSet;
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java:4]: Don't forget to remove unused imports.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 5 above)
Comment (pkuwm) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:158]: This is within a while loop. How many times this check will be run?

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 6 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:154]: It is very hard to explain. I will do my best to explain it here. (If it is still not clear, I suggest you to just run the test I added to this PR and it will be very clear to you.)
Explanation: 
Here is the _readyJobList contents after the controller switch in my test (without this fix):

> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB4]
> [testQueueParallelJobs_JOB5]
> [testQueueParallelJobs_JOB6]
> 

We end up having jut JOB4 in the DAG and the reason is because once each job is finished, we add new job to the _readyList (_readyList is updated in finishJob method). If there is only one Job in the _readyList, we will wait until 4 is finished and add JOB5 and so on. In this case, we will not run the whole queue in parallel. 

Also interesting to note here that even before controller swicth here is _readyList which is also wrong and we see two JOBs with the same name in the queue. However, this is not as critical as the the one I explained above:

> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB1]
> [testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB2]
> [testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB3]
> [testQueueParallelJobs_JOB3]

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 2 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:157]: Sure. I will add that now.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java (hunk 7 above)
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestEnqueueJobs.java]: Done.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 5 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java:158]: Usually in normal cases where we do not have controller switch we don't run this loop multiple times because _lastJob is keeping us at the end for the list. Also if we have controller switch, it will help us in the long run because we do not need to loop through the jobs that has been scheduled before multiple times.

File: helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java (hunk 6 above)
Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java]: Please note that if we have a lot of jobs in the queue and restart the controller, if we have for example 1000 jobs already in the queue and completed, how many times we have to loop through each job for the first pipeline. It will make the controller very slow (for the fist pipeline) because each job will go to _readyList multiple times.

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java]: With this fix, this is the _readyList after controller switch:
[testQueueParallelJobs_JOB0, testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2]
[testQueueParallelJobs_JOB1, testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3]
[testQueueParallelJobs_JOB2, testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4]
[testQueueParallelJobs_JOB3, testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB5]
[testQueueParallelJobs_JOB4, testQueueParallelJobs_JOB5, testQueueParallelJobs_JOB6]
[testQueueParallelJobs_JOB5, testQueueParallelJobs_JOB6]
[testQueueParallelJobs_JOB6]

Comment (alirezazamani) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java]: Without the fix:
Please note the we add job to the _readyJob list once we call finishJob (once a job is completed).
Let's say controller switch happened and we started from scratch.
We have [Job0, Job1 , Job2] in the _readyList.
To schedule the jobs, we get next job from readyList (Job0), see that this job is completed, we add the children of Job0 which is Job1. Hence the _readyList will be [Job1, Job2, Job1]. That is why we have same job added to the _readyList.

In conclusion, if a job is finished and controller adds a wrong job, the _readyJobList will fall behind. For example in my test, we add Job4 multiple times. Since Job4 goes to in-progress state and it takes some time to finish, we do not add new job (i.e. job5) until Job4 is finished and hence we do not run jobs in parallel if this happens.

Comment (narendly) [reply helix-core/src/main/java/org/apache/helix/task/RuntimeJobDag.java]: @alirezazamani Thank for explaining with concrete examples. So the root cause seems to be that we do not do if (!_readyJobList.contains(nextJob)) when we add the jobs to the ready list.

Good find and good fix! :)

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @dasahcc and @narendly 

Fix adding same job multiple times to RuntimeJobDag 

In this commit, a fix has been implemented to avoid
 _readyJobList in RuntimeJobDag to contain multiple
entries of the same job.
